src/Components/WebAssembly/Server/src/ContentEncodingNegotiator.cs

```diff
@@ -0,0 +1,123 @@
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Server
+{
+    internal class ContentEncodingNegotiator
+    {
+        // List of encodings by preference order with their associated extension so that we can easily handle "*".
+        private static readonly StringSegment[] _preferredEncodings =
+            new StringSegment[] { "br", "gzip" };
+
+        private static readonly Dictionary<StringSegment, string> _encodingExtensionMap = new Dictionary<StringSegment, string>(StringSegmentComparer.OrdinalIgnoreCase)
+        {
+            ["br"] = ".br",
+            ["gzip"] = ".gz"
+        };
+
+        private readonly RequestDelegate _next;
+        private readonly IWebHostEnvironment _webHostEnvironment;
+
+        public ContentEncodingNegotiator(RequestDelegate next, IWebHostEnvironment webHostEnvironment)
+        {
+            _next = next;
+            _webHostEnvironment = webHostEnvironment;
+        }
+
+        public Task InvokeAsync(HttpContext context)
+        {
+            NegotiateEncoding(context);
+            return _next(context);
+        }
+
+        private void NegotiateEncoding(HttpContext context)
+        {
+            var accept = context.Request.Headers[HeaderNames.AcceptEncoding];
+
+            if (StringValues.IsNullOrEmpty(accept))
+            {
+                return;
+            }
+
+            if (!StringWithQualityHeaderValue.TryParseList(accept, out var encodings) || encodings.Count == 0)
+            {
+                return;
+            }
+
+            var selectedEncoding = StringSegment.Empty;
+            var selectedEncodingQuality = .0;
+
+            foreach (var encoding in encodings)
+            {
+                var encodingName = encoding.Value;
+                var quality = encoding.Quality.GetValueOrDefault(1);
+
+                if (quality >= double.Epsilon && quality >= selectedEncodingQuality)
+                {
+                    if (quality == selectedEncodingQuality)
+                    {
+                        selectedEncoding = PickPreferredEncoding(context, selectedEncoding, encoding);
+                    }
+                    else if (_encodingExtensionMap.TryGetValue(encodingName, out var encodingExtension) && ResourceExists(context, encodingExtension))
+                    {
+                        selectedEncoding = encodingName;
+                        selectedEncodingQuality = quality;
+                    }
+
+                    if (StringSegment.Equals("*", encodingName, StringComparison.Ordinal))
+                    {
+                        // If we *, pick the first preferrent encoding for which a resource exists.
+                        selectedEncoding = PickPreferredEncoding(context, _preferredEncodings[^1], encoding);
```

Comment:
```suggestion
                        selectedEncoding = PickPreferredEncoding(context, _preferredEncodings[0], encoding);
```

Looks like `_preferredEncodings` is ordered with preferred first. We do prefer brotli, right?

Reply:
Also don't we need to loop over all the preferred encodings and pick the first one corresponding to a file on disk? Looks like currently it only considers whether a single item matches a file on disk because of the implementation inside `PickPreferredEncoding`.

Reply:
If we are evaluating "*" we need to iterate potentially from beginning to end until we find the one that we prefer the most that exists on disk, which is what `[^1]`